controllers/tlspolicy_certmanager_certificates.go

```diff
@@ -31,7 +32,9 @@ func (r *TLSPolicyReconciler) reconcileCertificates(ctx context.Context, tlsPoli
 	}
 
 	// Reconcile Certificates for each gateway directly referred by the policy (existing and new)
-	for _, gw := range append(gwDiffObj.GatewaysWithValidPolicyRef, gwDiffObj.GatewaysMissingPolicyRef...) {
+	gwList := append(gwDiffObj.GatewaysWithValidPolicyRef, gwDiffObj.GatewaysMissingPolicyRef...)
+	time.Sleep(250 * time.Millisecond) // Sleep required to make "should delete tls certificate when listener is removed" integration test pass.
```

Comment:
While I do agree with you that we should not be adding the sleep, after two days of probing around I could not find any thing better than this. The whole thing has something fishy about it. If you flip the list creation after the sleep the test will fail. It makes no sense and at some point the work needs to be done. Personal I would prefer to have the sleep in the code over the having the bug that the PR fixes.

Reply:
@Boomatang Do you have more details on why this test in particular is an issue? Does the test fail consistently? How does it fail (What error)?

Reply:
Had a look into this. This seems to be something to do more with the test scenario. The created test gateway status has the following:
```
conditions:
  - lastTransitionTime: "1970-01-01T00:00:00Z"
    message: Waiting for controller
    reason: Pending
    status: Unknown
    type: Accepted
  - lastTransitionTime: "1970-01-01T00:00:00Z"
    message: Waiting for controller
    reason: Pending
    status: Unknown
    type: Programmed

```
And it never goes to a `Programmed: True` state.

The topology filters outs Gateways not in the `Programmed: true` state at:
https://github.com/Boomatang/kuadrant-operator/blob/dc5f92b6520bc9ee03740457ba512ce65381a5dd/pkg/library/gatewayapi/topology.go#L202-L206

This causes the change event at https://github.com/Boomatang/kuadrant-operator/blob/fix/GH614/controllers/tlspolicy_controller_test.go#L410-L413 to never be reconciled by the TLS Policy controller as the request was filtered out at the event mapper in
https://github.com/Boomatang/kuadrant-operator/blob/dc5f92b6520bc9ee03740457ba512ce65381a5dd/pkg/library/mappers/gateway.go#L59-L63

Reply:
While I don't think that was the issue. However with the latest changes on main this issue seems to have being resolved. So I have reverted the change.

Reply:
The TLS tests don't use a real Gateway class since it was never intended that the tests should require a gateway in a programmed state in order to work. When the TLS Policy was first added it was explicitly ignoring the programmed check https://github.com/Kuadrant/kuadrant-operator/commit/e51ddaf0a823919f1f49c11eaf5dd38fb4b57071#diff-38477e498c938e91320e277fd9be8bb1abfcbb26abc2af86c309865d49f19a3cL256-L277. Seems quite a bit has been changed in here since i last looked at it, i cant really work out how any of the TLS tests are working tbh as it appears we removed that some time ago.

Reply:
+1 to what @mikenairn said, so while the test is passing for now I think it will be flaky at best (at least for me). @Boomatang Let me know if you want to fix the regression issue as part of this PR, if not, I can look into that separately since I introduced that regression issue 🧑‍💻

Reply:
As this it's self is fixing a bug and has being opened for awhile, I think it would be best to try get to the bottom of this flakiness in a second PR.

Reply:
I see this got merged without fixing the flaky test it was introducing https://github.com/Kuadrant/kuadrant-operator/actions/runs/9699428897/job/26768361658?pr=716

Any plans to fix that?

Reply:
I plan on addressing it as part of https://github.com/Kuadrant/kuadrant-operator/pull/715